File: python-version/file_transfer/codec.py
from __future__ import annotations

from .errors import FileTransferError
from .format import DecodedFile, build_stream, parse_stream
from .png_codec import decode_png_to_rgb, encode_rgb_png


def encode_bytes(data: bytes, filename: str) -> tuple[bytes, dict[str, object]]:
    side, stream, metadata = build_stream(data, filename)
    return encode_rgb_png(side, side, stream), metadata


def decode_bytes(png: bytes) -> DecodedFile:
    width, height, rgb = decode_png_to_rgb(png)
    return parse_stream(rgb, width, height)


def inspect_bytes(png: bytes) -> dict[str, object]:
    decoded = decode_bytes(png)
    return decoded.metadata()


def encode_v2_bytes(data: bytes, filename: str) -> tuple[bytes, dict[str, object]]:
    try:
        from .v2_image import encode_v2_jpeg
    except ModuleNotFoundError as exc:
        if exc.name == "PIL":
            raise FileTransferError("MISSING_DEPENDENCY", "JPEG v2 需要 Pillow；请先运行 setup 安装依赖") from exc
        raise

    return encode_v2_jpeg(data, filename)


def decode_v2_bytes(image: bytes):
    try:
        from .v2_image import decode_v2_image
    except ModuleNotFoundError as exc:
        if exc.name == "PIL":
            raise FileTransferError("MISSING_DEPENDENCY", "JPEG v2 需要 Pillow；请先运行 setup 安装依赖") from exc
        raise

    return decode_v2_image(image)


def inspect_v2_bytes(image: bytes) -> dict[str, object]:
    return decode_v2_bytes(image).metadata()

File: node-version/src/codec.js
const { buildStream, parseStream } = require('./format');
const { decodePngToRgb, encodeRgbPng } = require('./png-codec');

function encodeBytes(data, filename) {
  const { side, stream, metadata } = buildStream(data, filename);
  return { png: encodeRgbPng(side, side, stream), metadata };
}

function decodeBytes(png) {
  const { width, height, rgb } = decodePngToRgb(png);
  return parseStream(rgb, width, height);
}

function inspectBytes(png) { return decodeBytes(png).metadata(); }

async function encodeV2Bytes(data, filename) {
  return require('./v2-image').encodeV2Jpeg(data, filename);
}

async function decodeV2Bytes(image) {
  return require('./v2-image').decodeV2Image(image);
}

async function inspectV2Bytes(image) {
  return (await decodeV2Bytes(image)).metadata();
}

module.exports = { encodeBytes, decodeBytes, inspectBytes, encodeV2Bytes, decodeV2Bytes, inspectV2Bytes };
